juju/juju
# improve checks before insert of instanceCharmProfileData and improve/simplify instanceCharmProfileDataWatcher

## Description of change

Improve checks before adding or removing an instanceCharmProfileData to avoid upgrade-charm failures due to excessive contention errors.  

Ensure that no instanceCharmProfileData docs were removed by the uniter before the UpgradeCharmProfileComplete is in a terminal state, e.g. any lxd-profile upgrade has not been required, completed successfully or completed with an error.

Simplify the instanceCharmProfileDataWatcher to reduce database queries and api calls

## QA steps
```
charm pull ubuntu
cd $GOPATH/github.com/juju/juju
juju bootstrap localhost --build-agent --config logging-config="<root>=DEBUG;unit=DEBUG
juju deploy ~/ubuntu ; juju deploy neutron-openvswitch-252 
juju add-relation neutron-openvswitch ubuntu 
juju deploy ./testcharms/charm-repo/quantal/lxd-profile-alt --to lxd 
juju deploy ./testcharms/charm-repo/quantal/lxd-profile 
juju deploy ./testcharms/charm-repo/quantal/lxd-profile-subordinate
juju add-relation lxd-profile-subordinate lxd-profile 
juju add-relation lxd-profile-subordinate ubuntu
juju deploy percona-cluster-270
```
Commands to test; check 'lxc profile list' to see profiles being added and removed, matching the charm rev number;  :
`juju upgrade-charm lxd-profile-subordinate --path ./testcharms/charm-repo/quantal/lxd-profile-subordinate`  <-- run multiple times
`juju upgrade-charm ubuntu --path ~/ubuntu`  <-- run multiple times
`juju upgrade-charm lxd-profile --path ./testcharms/charm-repo/quantal/lxd-profile`  <-- run multiple times
`juju upgrade-charm neutron-opensvswitch`  <-- charm rev 255 should install an lxd-profile
`juju add-unit ubuntu`
`juju upgrade-charm percona-cluster`
`juju add-unit lxd-profile-alt --to <machine with percona-cluster>`  <-- check that profile added to machine
`juju remove-unit lxd-profile-alt/<unit added to percona-cluster>` <-- check that profile removed from machine.

Test break profile and fix it:
`cp ./testcharms/charm-repo/quantal/lxd-profile/lxd-profile.yaml ./testcharms/charm-repo/quantal/lxd-profile-alt/.`
`juju upgrade-charm lxd-profile-alt  --path ./testcharms/charm-repo/quantal/lxd-profile-alt`  <-- this will cause a machine error where the lxd-profile-alt/0 is deployed.
`juju add-unit lxd-profile-alt`  <-- machine should not be in an error state
`git checkout testcharms/charm-repo/quantal/lxd-profile-alt/lxd-profile.yaml`
`juju upgrade-charm lxd-profile-alt  --path ./testcharms/charm-repo/quantal/lxd-profile-alt`  <-- will resolve the machine error lxd-profile-alt/0 and cause one for lxd-profile-alt/1
`juju upgrade-charm lxd-profile-alt  --path ./testcharms/charm-repo/quantal/lxd-profile-alt`   <-- both units shouldn't have a machine error.
